github.com/cockroachdb/pebble/internal/metamorphic: TestMetaTwoInstance failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20052809?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20052809?buildTab=artifacts#meta) on refs/heads/master @ [46650aa8660c](https://github.com/cockroachdb/pebble/commits/46650aa8660c):


```
=== RUN   TestMetaTwoInstance
--- FAIL: TestMetaTwoInstance (6.70s)
panic: unable to parse line "iter30.Next(\"\") // [false] error initializing data block metadata: column 2 is out of range [0, 0)\n": strconv.Atoi: parsing "iter30.Next(\"\")": invalid syntax [recovered]
	panic: unable to parse line "iter30.Next(\"\") // [false] error initializing data block metadata: column 2 is out of range [0, 0)\n": strconv.Atoi: parsing "iter30.Next(\"\")": invalid syntax

goroutine 265 [running]:
testing.tRunner.func1.2({0xca6fa0, 0xc00011a0f0})
	GOROOT/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x35e
panic({0xca6fa0?, 0xc00011a0f0?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/pebble/metamorphic.extractOp({0xc000827963, 0x63})
	external/com_github_cockroachdb_pebble/metamorphic/history.go:223 +0x17e
github.com/cockroachdb/pebble/metamorphic.reorderHistory({0xc00032e000, 0x16c5, 0xc000044680?})
	external/com_github_cockroachdb_pebble/metamorphic/history.go:204 +0xa5
github.com/cockroachdb/pebble/metamorphic.RunAndCompare.func4.2(0xc00025c000)
	external/com_github_cockroachdb_pebble/metamorphic/meta.go:319 +0xdd
testing.tRunner(0xc00025c000, 0xc000228000)
	GOROOT/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 210
	GOROOT/src/testing/testing.go:1743 +0x390
=== RUN   TestMetaTwoInstance/compare
--- FAIL: TestMetaTwoInstance/compare (0.05s)
=== RUN   TestMetaTwoInstance/compare/random-012
--- FAIL: TestMetaTwoInstance/compare/random-012 (0.00s)
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20058128?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20058128?buildTab=artifacts#meta) on refs/heads/master @ [46650aa8660c](https://github.com/cockroachdb/pebble/commits/46650aa8660c):


```
=== RUN   TestMetaTwoInstance
--- FAIL: TestMetaTwoInstance (10.18s)
panic: unable to parse line "iter11.Next(\"\") // [false] error initializing data block metadata: column 2 is out of range [0, 0)\n": strconv.Atoi: parsing "iter11.Next(\"\")": invalid syntax [recovered]
	panic: unable to parse line "iter11.Next(\"\") // [false] error initializing data block metadata: column 2 is out of range [0, 0)\n": strconv.Atoi: parsing "iter11.Next(\"\")": invalid syntax

goroutine 254 [running]:
testing.tRunner.func1.2({0xca6fa0, 0xc000300060})
	GOROOT/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x35e
panic({0xca6fa0?, 0xc000300060?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/pebble/metamorphic.extractOp({0xc0006c6f47, 0x63})
	external/com_github_cockroachdb_pebble/metamorphic/history.go:223 +0x17e
github.com/cockroachdb/pebble/metamorphic.reorderHistory({0xc000410000, 0x1f80, 0xc000044800?})
	external/com_github_cockroachdb_pebble/metamorphic/history.go:204 +0xa5
github.com/cockroachdb/pebble/metamorphic.RunAndCompare.func4.2(0xc00013c340)
	external/com_github_cockroachdb_pebble/metamorphic/meta.go:319 +0xdd
testing.tRunner(0xc00013c340, 0xc000184080)
	GOROOT/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 202
	GOROOT/src/testing/testing.go:1743 +0x390
=== RUN   TestMetaTwoInstance/compare
--- FAIL: TestMetaTwoInstance/compare (0.05s)
=== RUN   TestMetaTwoInstance/compare/random-001
--- FAIL: TestMetaTwoInstance/compare/random-001 (0.00s)
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20062754?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/20062754?buildTab=artifacts#meta) on refs/heads/master @ [46650aa8660c](https://github.com/cockroachdb/pebble/commits/46650aa8660c):


```
          compression=Fastest
          filter_policy=rocksdb.BuiltinBloomFilter
          filter_type=table
          index_block_size=4096
          target_file_size=134217728
        
        [TestOptions]
          key_format=testkeys
          threads=7
          seed_efos=12242633951737656968
          use_excise=true
          use_jemalloc_size_classes=true
        
        ===== OPS =====
        /artifacts/meta/250630-040517.1531417703854/ops
        ===== HISTORY =====
        /artifacts/meta/250630-040517.1531417703854/random-008/history
        ...[1772 lines skipped]...
        db2.Delete("gusygkbawaa") // <nil> #1584
        batch13.Delete("jdgpmop") // <nil> #1585
        batch13.Set("inrhyv@1", "stghcflsisbnym") // <nil> #1589
        // 04:05:23.396 INFO: [JOB 29] ingesting: sstable created 000037
        // 04:05:23.396 INFO: [JOB 30] WAL created 000038
        // 04:05:23.396 INFO: [JOB 31] WAL created 000039
        iter15.Next("") // [true,"dixe",<no point>,["dixe","nmqqmuj")=>{""="ep"}*] <nil> #1586
        iter15.Last() // [true,"ybanf",<no point>,["ybanf","zrfojn")=>{""="ta","@1"="ywol"}*] <nil> #1587
        iter15.SetBounds("dukpowocfrd", "zotq@2") // <nil> #1588
        iter15.SeekPrefixGE("xbbhjggkp@2") // [false] <nil> #1590
        iter15.SeekLT("mbfdmyfty@4", "") // [true,"dukpowocfrd",<no point>,["dukpowocfrd","nmqqmuj")=>{""="ep"}*] <nil> #1592
        iter15.SeekPrefixGE("pybmtwgmiijv") // [true,"pybmtwgmiijv",<no point>,["pybmtwgmiijv","pybmtwgmiijv\x00")=>{""="ta"}*] <nil> #1593
        iter15.SetBounds("dukpowocfrd", "dukpowocfrd") // <nil> #1594
        iter15.Last() // [false] <nil> #1596
        iter15.Prev("") // [false] <nil> #1597
        iter15.Prev("deunsmzln") // [invalid] <nil> #1598
        iter15.SeekGE("gusygkbawaa", "") // [false] <nil> #1599
        iter15.SeekLT("dkjiprcy", "") // [false] <nil> #1601
        iter15.SetBounds("dukpowocfrd", "dukpowocfrd") // <nil> #1603
        iter15.SeekGE("dukpowocfrd", "") // [false] <nil> #1604
        iter15.Prev("") // [false] <nil> #1605
        iter15.Prev("") // [false] <nil> #1606
        iter15.Next("") // [false] <nil> #1607
        iter15.SeekPrefixGE("wqhid@2") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #1608
        iter16 = iter15.Clone(false, "dukpowocfrd", "dukpowocfrd", 2 /* key types */, "", "@1", false /* use L6 filters */, "@4" /* masking suffix */) // <nil> #1609
        iter15.Next("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #1610
        iter15.SeekPrefixGE("mkpdonqkgfoz@3") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #1614
        iter15.SetBounds("mbgjqftqxf@3", "yqhogjlrnn@5") // <nil> #1623
        iter15.SeekLT("yqhogjlrnn@5", "") // [true,"ybanf",<no point>,["ybanf","yqhogjlrnn@5")=>{""="ta","@1"="ywol"}*] <nil> #1624
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/250630-040517.1531417703854/random-008 --try-to-reduce -v
--- FAIL: TestMetaTwoInstance/execution/random-008 (0.19s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1751256317153938327 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/20062755?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/20062755?buildTab=artifacts#meta) on refs/heads/master @ [46650aa8660c](https://github.com/cockroachdb/pebble/commits/46650aa8660c):


```
=== RUN   TestMetaTwoInstance
--- FAIL: TestMetaTwoInstance (30.21s)
panic: unable to parse line "iter12.Next(\"\") // [false] error initializing data block metadata: column 2 is out of range [0, 0)\n": strconv.Atoi: parsing "iter12.Next(\"\")": invalid syntax [recovered]
	panic: unable to parse line "iter12.Next(\"\") // [false] error initializing data block metadata: column 2 is out of range [0, 0)\n": strconv.Atoi: parsing "iter12.Next(\"\")": invalid syntax

goroutine 434 [running]:
testing.tRunner.func1.2({0x1119d20, 0xc000050130})
	GOROOT/src/testing/testing.go:1632 +0x3fc
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x6b6
panic({0x1119d20?, 0xc000050130?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/pebble/metamorphic.extractOp({0xc0007473e1, 0x63})
	external/com_github_cockroachdb_pebble/metamorphic/history.go:223 +0x1d9
github.com/cockroachdb/pebble/metamorphic.reorderHistory({0xc0003b4000, 0xf92, 0xc000044380?})
	external/com_github_cockroachdb_pebble/metamorphic/history.go:204 +0xc7
github.com/cockroachdb/pebble/metamorphic.RunAndCompare.func4.2(0xc0003021a0)
	external/com_github_cockroachdb_pebble/metamorphic/meta.go:319 +0x145
testing.tRunner(0xc0003021a0, 0xc0000c4000)
	GOROOT/src/testing/testing.go:1690 +0x227
created by testing.(*T).Run in goroutine 365
	GOROOT/src/testing/testing.go:1743 +0x826
=== RUN   TestMetaTwoInstance/compare
--- FAIL: TestMetaTwoInstance/compare (0.23s)
=== RUN   TestMetaTwoInstance/compare/random-016
--- FAIL: TestMetaTwoInstance/compare/random-016 (0.00s)
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/sstable/reader_iter_single_lvl.go
+++ b/sstable/reader_iter_single_lvl.go
@@ -165,9 +165,6 @@ type singleLevelIterator[I any, PI indexBlockIterator[I], D any, PD dataBlockIte
 	useFilterBlock         bool
 	lastBloomFilterMatched bool
 
-	// Lazy loading flag
-	indexLoaded bool
-
 	transforms IterTransforms
 
 	// All fields above this field are cleared when resetting the iterator for reuse.
@@ -217,9 +214,14 @@ func newColumnBlockSingleLevelIterator(
 		i.vbRH = r.blockReader.UsePreallocatedReadHandle(objstorage.NoReadBefore, &i.vbRHPrealloc)
 	}
 	i.data.InitOnce(r.keySchema, r.Comparer, &i.internalValueConstructor)
-
-	// Use lazy loading by default - index will be loaded on first access
-	i.indexLoaded = false
+	indexH, err := r.readTopLevelIndexBlock(ctx, i.readEnv.Block, i.indexFilterRH)
+	if err == nil {
+		err = i.index.InitHandle(r.Comparer, indexH, opts.Transforms)
+	}
+	if err != nil {
+		_ = i.Close()
+		return nil, err
+	}
 	return i, nil
 }
 
@@ -252,8 +254,14 @@ func newRowBlockSingleLevelIterator(
 		i.data.SetHasValuePrefix(true)
 	}
 
-	// Use lazy loading by default - index will be loaded on first access
-	i.indexLoaded = false
+	indexH, err := r.readTopLevelIndexBlock(ctx, i.readEnv.Block, i.indexFilterRH)
+	if err == nil {
+		err = i.index.InitHandle(r.Comparer, indexH, opts.Transforms)
+	}
+	if err != nil {
+		_ = i.Close()
+		return nil, err
+	}
 	return i, nil
 }
 
@@ -441,7 +449,7 @@ func (i *singleLevelIterator[I, PI, P, PD]) SetContext(ctx context.Context) {
 // unpositioned. If unsuccessful, it sets i.err to any error encountered, which
 // may be nil if we have simply exhausted the entire table.
 func (i *singleLevelIterator[I, PI, P, PD]) loadDataBlock(dir int8) loadBlockResult {
-	if i.err != nil || !PI(&i.index).Valid() {
+	if !PI(&i.index).Valid() {
 		// Ensure the data block iterator is invalidated even if loading of the block
 		// fails.
 		PD(&i.data).Invalidate()
@@ -513,12 +521,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) ReadValueBlock(
 // apprioriate bound, depending on the iteration direction, and returns either
 // `blockIntersects` or `blockExcluded`.
 func (i *singleLevelIterator[I, PI, D, PD]) resolveMaybeExcluded(dir int8) intersectsResult {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return blockExcluded
-		}
-	}
 	// TODO(jackson): We could first try comparing to top-level index block's
 	// key, and if within bounds avoid per-data block key comparisons.
 
@@ -680,13 +682,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) SeekGE(
 func (i *singleLevelIterator[I, PI, D, PD]) seekGEHelper(
 	key []byte, boundsCmp int, flags base.SeekGEFlags,
 ) *base.InternalKV {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
-
 	// Invariant: trySeekUsingNext => !i.data.isDataInvalidated() && i.exhaustedBounds != +1
 
 	// SeekGE performs various step-instead-of-seeking optimizations: eg enabled
@@ -823,7 +818,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) seekPrefixGE(
 			flags = flags.DisableTrySeekUsingNext()
 		}
 		i.lastBloomFilterMatched = false
-
 		// Check prefix bloom filter.
 		var mayContain bool
 		mayContain, i.err = i.bloomFilterMayContain(prefix)
@@ -928,13 +922,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) virtualLastSeekLE() *base.InternalKV
 	i.boundsCmp = 0
 	i.positionedUsingLatestBounds = true
 
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
-
 	indexOk := PI(&i.index).SeekGE(key)
 	// We can have multiple internal keys with the same user key as the seek
 	// key. In that case, we want the last (greatest) internal key.
@@ -1006,12 +993,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) virtualLastSeekLE() *base.InternalKV
 func (i *singleLevelIterator[I, PI, D, PD]) SeekLT(
 	key []byte, flags base.SeekLTFlags,
 ) *base.InternalKV {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
 	if i.readEnv.Virtual != nil {
 		// Might have to fix upper bound since virtual sstable bounds are not
 		// known to callers of SeekLT.
@@ -1135,12 +1116,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) First() *base.InternalKV {
 // index file. For the latter, one cannot make any claims about absolute
 // positioning.
 func (i *singleLevelIterator[I, PI, D, PD]) firstInternal() *base.InternalKV {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
 	i.exhaustedBounds = 0
 	i.err = nil // clear cached iteration error
 	// Seek optimization only applies until iterator is first positioned after SetBounds.
@@ -1205,12 +1180,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) Last() *base.InternalKV {
 // index file. For the latter, one cannot make any claims about absolute
 // positioning.
 func (i *singleLevelIterator[I, PI, D, PD]) lastInternal() *base.InternalKV {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
 	i.exhaustedBounds = 0
 	i.err = nil // clear cached iteration error
 	// Seek optimization only applies until iterator is first positioned after SetBounds.
@@ -1280,12 +1249,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) Next() *base.InternalKV {
 
 // NextPrefix implements (base.InternalIterator).NextPrefix.
 func (i *singleLevelIterator[I, PI, D, PD]) NextPrefix(succKey []byte) *base.InternalKV {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
 	if i.exhaustedBounds == +1 {
 		panic("NextPrefix called even though exhausted upper bound")
 	}
@@ -1380,12 +1343,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) Prev() *base.InternalKV {
 }
 
 func (i *singleLevelIterator[I, PI, D, PD]) skipForward() *base.InternalKV {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
 	for {
 		if !PI(&i.index).Next() {
 			PD(&i.data).Invalidate()
@@ -1464,12 +1421,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) skipForward() *base.InternalKV {
 }
 
 func (i *singleLevelIterator[I, PI, D, PD]) skipBackward() *base.InternalKV {
-	if !i.indexLoaded {
-		if err := i.ensureIndexLoaded(); err != nil {
-			i.err = err
-			return nil
-		}
-	}
 	for {
 		if !PI(&i.index).Prev() {
 			PD(&i.data).Invalidate()
@@ -1563,8 +1514,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) closeInternal() error {
 	}
 	var err error
 	err = firstError(err, PD(&i.data).Close())
-	// Always close index iterator unconditionally to avoid BufferPool panic
-	// Even if lazy loading wasn't used, the index might have been initialized
 	err = firstError(err, PI(&i.index).Close())
 	if i.indexFilterRH != nil {
 		err = firstError(err, i.indexFilterRH.Close())
@@ -1583,7 +1532,6 @@ func (i *singleLevelIterator[I, PI, D, PD]) closeInternal() error {
 		err = firstError(err, i.vbRH.Close())
 		i.vbRH = nil
 	}
-	i.indexLoaded = false
 	return err
 }
 
@@ -1597,17 +1545,4 @@ func (i *singleLevelIterator[I, PI, D, PD]) String() string {
 // DebugTree is part of the InternalIterator interface.
 func (i *singleLevelIterator[I, PI, D, PD]) DebugTree(tp treeprinter.Node) {
 	tp.Childf("%T(%p) fileNum=%s", i, i, i.String())
-}
-
-func (i *singleLevelIterator[I, PI, D, PD]) ensureIndexLoaded() error {
-	indexH, err := i.reader.readTopLevelIndexBlock(i.ctx, i.readEnv.Block, i.indexFilterRH)
-	if err == nil {
-		err = PI(&i.index).InitHandle(i.reader.Comparer, indexH, i.transforms)
-	}
-	if err != nil {
-		return err
-	}
-
-	i.indexLoaded = true
-	return nil
 }

--- a/sstable/reader_iter_test.go
+++ b/sstable/reader_iter_test.go
@@ -53,20 +53,15 @@ func TestIteratorErrorOnInit(t *testing.T) {
 	defer toggle.Off()
 
 	var stats base.InternalIteratorStats
-	for range 20 {
+	for k := 0; k < 20; k++ {
 		if rand.IntN(2) == 0 {
-			iter, err := newRowBlockSingleLevelIterator(context.Background(), r, IterOptions{
+			_, err := newRowBlockSingleLevelIterator(context.Background(), r, IterOptions{
 				Transforms:           NoTransforms,
 				FilterBlockSizeLimit: NeverUseFilterBlock,
 				Env:                  ReadEnv{Block: block.ReadEnv{Stats: &stats, BufferPool: &pool}},
 				ReaderProvider:       MakeTrivialReaderProvider(r),
 			})
-			// Single-level iterators use lazy loading - creation succeeds but first access fails
-			require.NoError(t, err)
-			// Error should surface when trying to use the iterator
-			_ = iter.First()
-			require.Error(t, iter.Error())
-			iter.Close()
+			require.Error(t, err)
 		} else {
 			_, err := newRowBlockTwoLevelIterator(context.Background(), r, IterOptions{
 				Transforms:           NoTransforms,

--- a/sstable/reader_iter_two_lvl.go
+++ b/sstable/reader_iter_two_lvl.go
@@ -78,7 +78,6 @@ func (i *twoLevelIterator[I, PI, D, PD]) loadSecondLevelIndexBlock(dir int8) loa
 		i.secondLevel.err = err
 		return loadBlockFailed
 	}
-	i.secondLevel.indexLoaded = true
 	return loadBlockOK
 }
 

--- a/testdata/checkpoint
+++ b/testdata/checkpoint
@@ -261,16 +261,16 @@ open: db/000007.sst (options: *vfs.randomReadsOption)
 read-at(700, 61): db/000007.sst
 read-at(649, 51): db/000007.sst
 read-at(132, 517): db/000007.sst
-open: db/000005.sst (options: *vfs.sequentialReadsOption)
 read-at(91, 41): db/000005.sst
+open: db/000005.sst (options: *vfs.sequentialReadsOption)
 read-at(0, 91): db/000005.sst
-open: db/000007.sst (options: *vfs.sequentialReadsOption)
 read-at(91, 41): db/000007.sst
+open: db/000007.sst (options: *vfs.sequentialReadsOption)
 read-at(0, 91): db/000007.sst
 create: db/000010.sst
 close: db/000005.sst
-open: db/000009.sst (options: *vfs.sequentialReadsOption)
 read-at(95, 41): db/000009.sst
+open: db/000009.sst (options: *vfs.sequentialReadsOption)
 read-at(0, 95): db/000009.sst
 close: db/000007.sst
 close: db/000009.sst

--- a/testdata/cleaner
+++ b/testdata/cleaner
@@ -82,13 +82,13 @@ open: db/000007.sst (options: *vfs.randomReadsOption)
 read-at(501, 61): db/000007.sst
 read-at(472, 37): db/000007.sst
 read-at(53, 419): db/000007.sst
-open: db/000005.sst (options: *vfs.sequentialReadsOption)
 read-at(52, 27): db/000005.sst
+open: db/000005.sst (options: *vfs.sequentialReadsOption)
 read-at(0, 52): db/000005.sst
 create: db/000008.sst
 close: db/000005.sst
-open: db/000007.sst (options: *vfs.sequentialReadsOption)
 read-at(26, 27): db/000007.sst
+open: db/000007.sst (options: *vfs.sequentialReadsOption)
 read-at(0, 26): db/000007.sst
 close: db/000007.sst
 sync-data: db/000008.sst

--- a/testdata/event_listener
+++ b/testdata/event_listener
@@ -145,11 +145,11 @@ open: db/000008.sst (options: *vfs.randomReadsOption)
 read-at(686, 61): db/000008.sst
 read-at(636, 50): db/000008.sst
 read-at(119, 517): db/000008.sst
-open: db/000005.sst (options: *vfs.sequentialReadsOption)
 read-at(78, 41): db/000005.sst
+open: db/000005.sst (options: *vfs.sequentialReadsOption)
 read-at(0, 78): db/000005.sst
-open: db/000008.sst (options: *vfs.sequentialReadsOption)
 read-at(78, 41): db/000008.sst
+open: db/000008.sst (options: *vfs.sequentialReadsOption)
 read-at(0, 78): db/000008.sst
 close: db/000008.sst
 close: db/000005.sst

--- a/testdata/flushable_ingest
+++ b/testdata/flushable_ingest
@@ -701,17 +701,17 @@ allowFlush
 get with-fs-logging
 small-00001
 ----
-read-at(199, 74): 000004.sst
 read-at(158, 41): 000004.sst
+read-at(199, 74): 000004.sst
 read-at(0, 158): 000004.sst
 small-00001:val-00001
 
-# When the key doesn't pass the bloom filter, we should see only one block
-# read due to lazy loading optimization - bloom filter is checked first,
-# and if it rejects the key, we can avoid loading the index block entirely.
+# When the key doesn't pass the bloom filter, we should see only two block
+# reads.
 get with-fs-logging
 small-00001-does-not-exist
 ----
+read-at(158, 41): 000004.sst
 read-at(199, 74): 000004.sst
 small-00001-does-not-exist: pebble: not found
 

